src/lib/updating-element.ts

```diff
@@ -0,0 +1,418 @@
+  // true to put property into observedAttributes
+  attribute?: boolean;
+  // defaults to lowercase
+  attributeName?: string;
+  // custom deserializer
+  fromAttribute?: Function;
+  // custom serializer
+  toAttribute?: Function;
+  // custom change function
+  shouldChange?: Function;
+}
+
+interface Properties {
+  [key: string]: PropertyOptions;
+}
+
+interface AttributeMap {
+  [key: string]: string;
+}
+
+interface AnyObject {
+  [key: string]: {};
+}
+
+/**
+ * Creates a property accessor on the given prototype if one does not exist.
+ * Uses `getProperty` and `setProperty` to manage the property's value.
+ */
+function makeProperty(name: string, proto: Object) {
+  if (proto.hasOwnProperty(name) || (name in proto)) {
+    return;
+  }
+  Object.defineProperty(proto, name, {
+    get() {
+      return this.getProperty(name);
+    },
+    set(value) {
+      this.setProperty(name, value);
+    },
+    configurable: true,
+    enumerable: true
+  });
+}
+
+/**
+ * Decorator which creates a property. Optionally a `PropertyOptions` object
+ * can be supplied with keys:
+ * * attribute: a boolean that should be true if the property should be an observedAttribute
+ * * fromAttribute: function used to deserialize an attribute value
+ * * toAttribute: function used to serialize a property to an attribute value
+ * * attributeName: string that is a custom attribute name (defaults to lowercased property name)
+ */
+export const property = (options?: PropertyOptions) => (proto: Object, name: string) => {
+  const ctor = proto.constructor as typeof UpdatingElement;
+  ctor._ensurePropertyStorage();
+  ctor._classProperties[name] = options || {};
+  makeProperty(name, proto);
+};
+
+/**
+ * Change function that returns true if `value` is different from `oldValue`.
+ * This method is used as the default for a property's `shouldChange` function.
+ */
+// tslint:disable-next-line no-any
+export function identity(value: any, old: any) {
+  // This ensures (old==NaN, value==NaN) always returns false
+  return old !== value && (old === old || value === value);
+}
+
+/**
+ * Base element class which manages element properties and attributes. When
+ * properties change, the `update` method is asynchronously called. This method
+ * should be supplied by subclassers to render updates as desired.
+ */
+export class UpdatingElement extends HTMLElement {
+
+  private _updatePromise: Promise<boolean>|null = null;
+  private _isEnabled = false;
+  private _isValidating = false;
+  private _hasUpdated = false;
+  private _isUpdating = false;
+  private _serializingProperty = false;
+  private _instanceProperties: AnyObject = {};
+  private _updateCompleteResolver: Function|null = null;
+
+  /**
+   * Node or ShadowRoot into which element DOM should be renderd. Defaults
+   * to an open shadowRoot.
+   */
+  root?: Element|DocumentFragment;
+
+  /**
+   * Object with keys for all properties with their current values.
+   */
+  props: AnyObject = {};
```

Comment:
I really hate the `props` name. Why not `properties`? Also, should this be public? Why wouldn't this be slightly hidden so that the accessors are the one public API for them?